### PR TITLE
Add concurrent writes reconciliation for non-blind INSERT in Delta Lake

### PR DESCRIPTION
**Pull request by findinpath**
## Description

Follow-up from https://github.com/trinodb/trino/pull/18506

Add concurrent writes reconciliation for non-blind INSERT in Delta Lake

On tables with isolation level WriteSerializable, allow committing
non-blind append INSERT operations in a concurrent context by
placing these operations right after any other previously concurrently
completed write operations.

For disambiguation, INSERT operations that contain subqueries reading the same
table are considered non-blind inserts.

Disallow committing insert operations in any of the following cases:

- table schema change has been committed in the meantime
- table protocol change has been committed in the meantime
- add files committed in the meantime should be read by
  the current insert operations
- remove files committed in the meantime conflict with the add files read
  by the current insert operations
  
  
 Relevant example taken from Databricks documentation in regards to the
 distinction between `WriteSerializable` and `Serializable` isolation levels:
 
 > For example, consider `txn1`, a long running delete and `txn2`,
 > which inserts blindly data into the table.
 > `txn2` and `txn1` complete and they are recorded in the order
 > `txn2, txn1`
 > into the history of the table.
 > According to the history, the data inserted in `txn2` should not exist
 > in the table. For `Serializable` level, a reader would never see data
 > inserted by `txn2`. However, for the `WriteSerializable` level, a reader
 > could at some point see the data inserted by `txn2`.
 
 
 A few words about WriteSerializable isolation level taken from delta.io javadocs:
 
 > This isolation level will ensure snapshot isolation consistency guarantee
 > between write operations only. 
 > In other words, if only the write operations are considered, then
 > there exists a serializable sequence between them that would produce the same
 > result as seen in the table.
 
 Consider the Delta Lake table `t` with the following definition:
 
 ```
 CREATE TABLE t(data integer);
 INSERT INTO t VALUES 0;
 ```
 
There will be always reconciliation when dealing with the following concurrent
 `INSERT` statements:

```
INSERT INTO t SELECT count(*) FROM t; -- non-blind insert
INSERT INTO t VALUES 2; -- blind insert
``` 

The expected state of the table `t` will be independently of the order of the way 
in which the operations were actually committed to the transaction log:

```
0
1
2
```
 
On the other hand, there can't be reconciliation when dealing with the following
 concurrent `INSERT` statements:

```
INSERT INTO t SELECT count(*) FROM t; -- non-blind insert
INSERT INTO t SELECT count(*) FROM t; -- non-blind insert
``` 

The reason why the reconciliation will fail is because there can't be picked any
serializable sequence of statements to explain the state of the table if both
statements would commit successfully.
Just for the sake of demonstration, if both statements would complete successfully,
the table state would be:

```
0
1
1
```   

However, independently how we would pick a serializable sequence of operations
performed on the table, the table state would look like:

```
0
1
2
``` 
 
 

## Additional context and related issues

https://docs.delta.io/latest/concurrency-control.html
https://docs.databricks.com/en/optimizations/isolation-level.html#isolation-levels

Delta Lake OSS Source Code

https://github.com/delta-io/delta/blob/28881f5710e7b0b0bf2d959e36a5b6ebd75c62ae/spark/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala

Delta Lake OSS Code source explained

https://books.japila.pl/delta-lake-internals/OptimisticTransactionImpl/#checkForConflicts

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Delta Lake
* Concurrent write reconciliation for non-blind append INSERT operation. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by ebyhr** [comment]
/test-with-secrets sha=e452a09cf3bf6726cd1461b197f0797c784d75c7

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeConnectorSmokeTest.java [hunk 1]
```diff
@@ -87,6 +92,7 @@
 import static java.util.concurrent.Executors.newFixedThreadPool;
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeConnectorSmokeTest.java:92]: Note to self - add in the commit comment.

```
Relevant example taken from Databricks documentation in regards to the
distinction between `WriteSerializable` and `Serializable` isolation levels:

> For example, consider `txn1`, a long running delete and `txn2`,
> which inserts blindly (that is, operations that blindly append data
. without reading any data) data deleted by `txn1`.
> `txn2` and `txn1` complete and they are recorded in the order
> `txn2, txn1`
> into the history of the table.
> According to the history, the data inserted in `txn2` should not exist
> in the table. For `Serializable` level, a reader would never see data
> inserted by `txn2`. However, for the `WriteSerializable` level, a reader
> could at some point see the data inserted by `txn2`.

```

Comment (findinpath) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeConnectorSmokeTest.java:92]: add as well https://github.com/delta-io/delta/blob/master/PROTOCOL.md#Action-Reconciliation

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 2]
```diff
@@ -1865,12 +1865,17 @@ private void checkForInsertTransactionConflicts(
...
 
-        if (!isBlindAppend) {
-            throw new TransactionFailedException("Conflicting concurrent writes with the current non blind append INSERT operation");
-        }
-
-        if (isolationLevel == SERIALIZABLE) {
-            throw new TransactionFailedException("Conflicting concurrent writes with the current blind append INSERT operation on Serializable isolation level");
+        switch (isolationLevel) {
+            case WRITESERIALIZABLE -> {
+                List<TupleDomain<DeltaLakeColumnHandle>> enforcedPartitionConstraints = sameAsTargetSourceTableHandles.stream()
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1870]: enforcedPartitionConstraints - change name

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 3]
```diff
@@ -1890,6 +1895,46 @@ private static void checkNoMetadataUpdates(DeltaLakeCommitSummary commitSummary)
         }
     }
 
+    private static void checkIfCommittedAddedFilesConflictWithCurrentOperation(TupleDomain<DeltaLakeColumnHandle> enforcedPartitionConstraint, IsolationLevel isolationLevel, DeltaLakeCommitSummary commitSummary)
+    {
+        List<AddFileEntry> addedFiles = commitSummary.getIsBlindAppend().orElse(false)
+                ? List.of() // Do not conflict with blind appends
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1901]: Explain the reasoning why .

Thank you @pajaks  for pointing out why do we actually conflict with non-blind inserts for WriteSerializable.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 4]
```diff
@@ -1890,6 +1895,46 @@ private static void checkNoMetadataUpdates(DeltaLakeCommitSummary commitSummary)
...
+                .findFirst();
+        if (unreadAddAction.isPresent()) {
+            throw new TransactionFailedException("Conflicting concurrent writes found. Data files were  added on the table with isolation level %s in the partition '%s' of the modified table  by concurrent write operation."
+                    .formatted(isolationLevel, unreadAddAction.get().getPartitionValues()));
+        }
+    }
+
+    private static void checkIfCommittedRemovedFilesConflictWithCurrentOperation(IsolationLevel isolationLevel, DeltaLakeCommitSummary commitSummary)
+    {
+        List<RemoveFileEntry> removedFiles = commitSummary.getIsBlindAppend().orElse(false)
```
Comment (pajaks) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1925]: Can we have blind append when there are remove files?
Maybe we should always check `removedFiles.isEmpty` to be sure?

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeConnectorSmokeTest.java [hunk 5]
```diff
@@ -2326,6 +2332,128 @@ private void testConcurrentInsertsReconciliationForBlindInserts(boolean partitio
...
+        }
+        finally {
+            assertUpdate("DROP TABLE " + tableName);
+            executor.shutdownNow();
+            assertTrue(executor.awaitTermination(10, SECONDS));
+        }
+    }
+
+    @RepeatedTest(3)
+    public void testConcurrentInsertsReconciliationForMixedInserts()
```
Comment (pajaks) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeConnectorSmokeTest.java:2404]: Can we have also test with UPDATE, REMOVE?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 6]
```diff
@@ -1890,6 +1895,46 @@ private static void checkNoMetadataUpdates(DeltaLakeCommitSummary commitSummary)
...
+        }
+
+        boolean readWholeTable = enforcedPartitionConstraint.isAll();
+        if (readWholeTable) {
+            throw new TransactionFailedException("Conflicting concurrent writes found. Data files added in the modified table by concurrent write operation.");
+        }
+
+        Map<DeltaLakeColumnHandle, Domain> enforcedDomains = enforcedPartitionConstraint.getDomains().orElseThrow();
+        Optional<AddFileEntry> unreadAddAction = addedFiles.stream()
+                .filter(addAction -> partitionMatchesPredicate(addAction.getCanonicalPartitionValues(), enforcedDomains))
```
Comment (pajaks) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1915]: Should it match source enforceDomains with source addAction. Or with target?

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeConnectorSmokeTest.java (hunk 5 above)
Comment (findinpath) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeConnectorSmokeTest.java:2404]: Unfortunately, not yet - this would be in a follow-up where `UPDATE` & `DELETE` are able to handle reconciliation.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java (hunk 6 above)
Comment (findinpath) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1915]: The reconciliation is done AFAIU based on the enforced partition constraints of the source tables (for non-blind INSERT or UPDATE/MERGE/DELETE operations).
If the operations writes new data - lNSERT operations - , this is not relevant for the WriteSerializable isolation level.
This aspect becomes relevant when dealing with Serializable operations on the same partition of the table - out of scope for the current PR.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeLocalConcurrentWritesTest.java [hunk 7]
```diff
@@ -167,6 +169,184 @@ private void testConcurrentInsertsReconciliationForBlindInserts(boolean partitio
...
+        }
+        finally {
+            assertUpdate("DROP TABLE " + tableName);
+            executor.shutdownNow();
+            assertTrue(executor.awaitTermination(10, SECONDS));
+        }
+    }
+
+    @Test
+    public void testConcurrentInsertsSelectingFromDifferentPartitionsOfSameTable()
```
Comment (pajaks) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeLocalConcurrentWritesTest.java:298]: Any reason why this test is not in `BaseDeltaLakeConnectorSmokeTest` as well?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 8]
```diff
@@ -1890,6 +1893,54 @@ private static void checkNoMetadataUpdates(DeltaLakeCommitSummary commitSummary)
...
+                .findFirst();
+        if (unreadAddAction.isPresent()) {
+            throw new TransactionFailedException("Conflicting concurrent writes found. Data files were  added on the table with isolation level %s in the partition '%s' of the modified table  by concurrent write operation."
+                    .formatted(isolationLevel, unreadAddAction.get().getPartitionValues()));
+        }
+    }
+
+    private static void checkIfCommittedRemovedFilesConflictWithCurrentOperation(IsolationLevel isolationLevel, DeltaLakeCommitSummary commitSummary)
+    {
+        if (commitSummary.getIsBlindAppend().orElse(false)) {
```
Comment (pajaks) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1925]: After some more testing I found that concurrent `OPTIMIZE` is marked as blindAppend and have removedFiles in commitSummary, so it will fail. Should we mark it as not blindAppend?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 9]
```diff
@@ -1890,6 +1893,54 @@ private static void checkNoMetadataUpdates(DeltaLakeCommitSummary commitSummary)
...
+            return;
+        }
+
+        boolean readWholeTable = enforcedSourcePartitionConstraints.isAll();
+        if (readWholeTable) {
+            throw new TransactionFailedException("Conflicting concurrent writes found. Data files added in the modified table by concurrent write operation.");
+        }
+
+        Map<DeltaLakeColumnHandle, Domain> enforcedDomains = enforcedSourcePartitionConstraints.getDomains().orElseThrow();
+        Optional<AddFileEntry> unreadAddAction = addedFiles.stream()
```
[the remark this answers is not in the document]
Comment (findinpath) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1914]: The files are not necessarily conflicting.
It may very well be that the operations don't overlap at row level.
Here we're just aborting the operation if we're findining new active files that should have been taken into account

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java (hunk 8 above)
Comment (findinpath) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1925]: Correct. The `OPTIMIZE` needs to read from the table it optimizes :)

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeLocalConcurrentWritesTest.java (hunk 7 above)
Comment (findinpath) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeLocalConcurrentWritesTest.java:298]: I'm reluctant that there is added value to test extensively on  `BaseDeltaLakeConnectorSmokeTest`.
The class would not be a "smoke" test class anymore.

I trust that the `local` file store does indeed throw `FileAlreadyExistsException` when a concurrent operation creates a namesake transaction log file and that's why I'd be inclined to say `TestDeltaLakeLocalConcurrentWritesTest` is fit to hold an increased battery of tests for concurrent writes reconciliation functionality.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeConnectorSmokeTest.java [hunk 10]
```diff
@@ -2326,6 +2332,128 @@ private void testConcurrentInsertsReconciliationForBlindInserts(boolean partitio
...
+        assertUpdate("CREATE TABLE " + tableName + " (a, part) WITH (partitioned_by = ARRAY['part']) AS VALUES (0, 10)", 1);
+
+        try {
+            // insert data concurrently
+            executor.invokeAll(ImmutableList.<Callable<Void>>builder()
+                            .add(() -> {
+                                barrier.await(10, SECONDS);
+                                // Independent of the execution order, use on the non-blind insert a partition constraint which is not `TupleDomain.all()`
+                                // which should lead to reconciliation
+                                getQueryRunner().execute("INSERT INTO " + tableName + " SELECT COUNT(*) AS a, part FROM " + tableName + " WHERE part = 10 GROUP BY part");
```
[the remark this answers is not in the document]
Comment (findinpath) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeConnectorSmokeTest.java:2420]: I modified the test to contain two competing non-blind `INSERT` operations

**Review by pajaks — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeCommitSummary.java [hunk 11]
```diff
@@ -64,6 +87,21 @@ public Optional<ProtocolEntry> getProtocol()
         return protocol;
     }
 
+    public Optional<CommitInfoEntry> getCommitInfo()
+    {
+        return commitInfo;
+    }
+
+    public List<RemoveFileEntry> getRemovedFiles()
```
[the remark this answers is not in the document]
Comment (findinpath) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeCommitSummary.java:95]: See `io.trino.plugin.deltalake.DeltaLakeMetadata#checkIfCommittedRemovedFilesConflictWithCurrentOperation`

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 12]
```diff
@@ -1908,6 +1912,54 @@ private static void checkNoMetadataUpdates(DeltaLakeCommitSummary commitSummary)
...
+                .filter(addAction -> partitionMatchesPredicate(addAction.getCanonicalPartitionValues(), enforcedDomains))
+                .findFirst();
```
[the remark this answers is not in the document]
Comment (findinpath) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1935]: Probably exposing which partitions are actually conflicting is not necessarily of help for the customer.
I removed the logic collecting the conflicting first partition.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeCommitSummary.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeCommitSummary.java:90]: 👍 Sorry, I left a comment in the wrong place.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeCommitSummary.java [hunk 13]
```diff
@@ -26,16 +28,22 @@
 
 public class DeltaLakeCommitSummary
 {
+    private final long version;
     private final List<MetadataEntry> metadataUpdates;
     private final Optional<ProtocolEntry> protocol;
+    private final Optional<CommitInfoEntry> commitInfo;
+    private final List<RemoveFileEntry> removedFiles;
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeCommitSummary.java:35]: Could you summarize the data in this summary more? Looks like you could throw out any add file commits that are blind appends entirely, and for other commits you need a set of the partitions but not the rest of the data.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 14]
```diff
@@ -1908,6 +1912,48 @@ private static void checkNoMetadataUpdates(DeltaLakeCommitSummary commitSummary)
...
+            // when backtracking which serializable sequence of operations led to the current state of the table.
+            checkState(commitSummary.getRemovedFiles().isEmpty(), "Blind append transaction %s cannot contain removed files", commitSummary.getVersion());
+            return;
+        }
+
+        if (commitSummary.getRemovedFiles().isEmpty()) {
+            return;
+        }
+
+        // TODO Pass active files of the target table read in DeltaLakeSplitSource to figure out whether the removed files do actually conflict with the read table partitions
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1953]: Is it the read partitions that matter for conflicts here or the written partitions?

**Review by alexjo2144 — COMMENTED** [review]
From what I can tell the conflict handling here is still safe, just a few questions/comments.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeCommitSummary.java (hunk 13 above)
Comment (findinpath) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeCommitSummary.java]: Keeping only:
- canonical partition keys for the added files
- a flag stating whether or not the commit contains remove files.

**Comment by findinpath** [comment]
@ebyhr , @alexjo2144 AC.
Please make another round of review 🙏

**Comment by ebyhr** [comment]
/test-with-secrets sha=4d0024c45c16cde39d18dc160e6eef83abcd619e

**Review by ebyhr — APPROVED** [review]

**Comment by ebyhr** [comment]
@findinpath  Could you take a look at the following failure? It's happening on master:
```
Error:    TestDeltaLakeLocalConcurrentWritesTest.testConcurrentInsertsSelectingFromTheSameTable:177->testConcurrentInsertsSelectingFromTheSameTable:228->AbstractTestQueryFramework.assertQuery:350 For query 20240329_103745_00065_e4xfv: 
 SELECT * FROM test_concurrent_inserts_select_from_same_table_h28nk6v7lf
not equal
Actual rows (up to 100 of 1 extra rows shown, 3 rows in total):
    [1, 10]
Expected rows (up to 100 of 1 missing rows shown, 3 rows in total):
    [2, 10]
```
https://github.com/trinodb/trino/actions/runs/8480015101/job/23235027443
